### PR TITLE
Remove location fields in SQLUserData

### PR DESCRIPTION
**Pull request by jingcheng16**
## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-15717

### Safety story

We used to store location info in two places: on `CouchUser` model and in `SQLUserData`. In previous PR, we have stopped writing location info to and reading location info from `SQLUserData`. This PR is just a follow up step to clean up some data we no longer needed.

I tested applying the migration and reverting it locally and on staging

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

Past code changes that are coordinated with this migration: https://github.com/dimagi/commcare-hq/pull/34852.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+def revert_keys_in_data(apps, schema_editor):
+    SQLUserData = apps.get_model('users', 'SQLUserData')
+    CommCareUserCouch = CommCareUser.get_db()
+
+    for user_data in SQLUserData.objects.all():
+        user_id = user_data.user_id
+
+        try:
+            commcare_user = CommCareUserCouch.get(user_id)
+        except ResourceNotFound:
```
Comment (jingcheng16) [inline corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:38]: When I use `SQLUserData`'s `user_id` to query `Couch` doc, I can get `ResourceNotFound: deleted`. Does it mean that user has been deleted? And should we delete the related `SQLUserData` object as well?

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        keys_to_remove = ['commcare_location_id', 'commcare_location_ids', 'commcare_primary_case_sharing_id']
+
+        modified = False
+        for key in keys_to_remove:
+            if key in data:
+                del data[key]
+                modified = True
+
+        if modified:
+            user_data.data = data
```
Comment (millerdev) [inline corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:25]: nit: I think this reassignment is unnecessary since `user_data.data` is the same dict as `data`.
```suggestion
```

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        location_id = commcare_user.get('location_id')
+        assigned_location_ids = commcare_user.get('assigned_location_ids', [])
+
+        if location_id:
+            data['commcare_location_id'] = location_id
+            data['commcare_primary_case_sharing_id'] = location_id
+        if assigned_location_ids:
+            data['commcare_location_ids'] = user_location_data(assigned_location_ids)
+
+        user_data.data = data
```
Comment (millerdev) [inline corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:51]: Same as above
```suggestion
```

**Review by millerdev — APPROVED** [review]

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:38]: I remember deletions being a little weird when working on this.  `SQLUserData` is tied to the Django `User` for a given web user, and should be deleted automatically when that is deleted.  This sounds like a couch user was deleted, but not the corresponding Django user (and user data).  Not sure what could cause that.

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+from corehq.apps.users.util import user_location_data
+from corehq.apps.users.models import CommCareUser
+from corehq.util.django_migrations import skip_on_fresh_install
+
+
+@skip_on_fresh_install
+def remove_keys_from_data(apps, schema_editor):
+    SQLUserData = apps.get_model('users', 'SQLUserData')
+
+    for user_data in SQLUserData.objects.all():
```
Comment (esoergel) [inline corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: You might want to change this to make it so you can run the migration before deploy.  I think there are two options:
1. Add a filter to this queryset to include only rows that have one or more of the keys present (`user_data.data` is a `JSONField`, and so should be queryable)
2. Add a `@once_off_migration` decorator somewhere.

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            if key in data:
+                del data[key]
+                modified = True
+
+        if modified:
+            user_data.data = data
+            user_data.save()
+
+
+def revert_keys_in_data(apps, schema_editor):
```
Comment (esoergel) [inline corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:29]: We don't ever actually run reverse migrations, so I'm of the opinion that we shouldn't write them.  If there were a problem and we needed to undo this change, we'd write a new migration with this code in it and deploy that.  I'm certainly not opposed to preparing for the worst, just want to be explicit about expectations.

Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:29]: https://github.com/dimagi/commcare-hq/pull/34852#issuecomment-2252993403
^ though this is a decent enough reason to keep it around

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 4 above)
Comment (jingcheng16) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: Hi Ethan, why I need to change it to run the migration before deploy? I plan to run this after the PR to read from system data is merged.

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 5 above)
Comment (jingcheng16) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:29]: Thank you Ethan. I always thought in general we want to write reverse functions for migration. Is there any harm to leave the reverse function in a migration file?

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: We usually try to run migrations before the deploy if it might take a while, so as not to complicate the deploy itself.  Are you planning to deploy a private release and run the migration from there?  That could also work.

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 5 above)
Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:28]: nope, no harm leaving it in

Comment (jingcheng16) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:28]: Thank you! I'll just leave it here.

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 4 above)
Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: That should be fine then.  Sometimes it's preferred to run the actual data migration in a management command, and then have a django migration that calls that command to ensure it gets run on deploy later.  For that approach to work, you need something to either skip the already-migrated rows or skip the whole migration.  However, if you'll be running it via the django migration, then the migration framework essentially does that anyways, so no special handling is required here.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py (hunk 4 above)
Comment (jingcheng16) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: Thank you for your explanation! I will run migration before I merge the PR, so I can avoid blocking production deployment.

Comment (esoergel) [reply corehq/apps/users/migrations/0072_remove_location_fields_in_user_data.py:14]: Actually I think it's a best practice to merge before running the migration, that way all migrations on production are at least on `master`.  This is also a failsafe to ensure no one else merges a conflicting migration in the interim, which would be a mess to resolve.

There's some applicable information in this doc if helpful: https://commcare-hq.readthedocs.io/migrations_in_practice.html#
(Though I've been meaning to revisit/rewrite it)
And this doc describes the migration command pattern: https://commcare-hq.readthedocs.io/migration_command_pattern.html

**Comment by jingcheng16** [comment]
Close because I reworked on it in #35298
